File: algorithms/2_medium/C/00039_combination_sum/backtrack.py
class Solution:
    def backtrack(self, res, tmp, candidates, candidates_len, target, curr_index):
        if target == 0:
            res.append(tmp[:])
            return

        for i in  range(curr_index, candidates_len):
            if target - candidates[i] >= 0:
                tmp.append(candidates[i])
                self.backtrack(res, tmp, candidates, candidates_len, target - candidates[i], i)
                tmp.pop()
            else:
                break

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        tmp = []

        candidates.sort()
        self.backtrack(res, tmp, candidates, len(candidates), target, 0)

        return res

File: algorithms/2_medium/C/00039_combination_sum/backtrack.rs
impl Solution {
    fn backtrack(
        res: &mut Vec<Vec<i32>>,
        tmp: &mut Vec<i32>,
        candidates: &Vec<i32>,
        candidates_len: usize,
        curr_index: usize,
        target: i32,
    ) {
        if target == 0 {
            res.push(tmp.clone());
            return;
        }

        for i in curr_index..candidates_len {
            if target - candidates[i] >= 0 {
                tmp.push(candidates[i]);
                Self::backtrack(res, tmp, candidates, candidates_len, i, target - candidates[i]);
                tmp.pop();
            } else {
                break;
            }
        }
    }

    pub fn combination_sum(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        let mut tmp: Vec<i32> = Vec::new();

        candidates.sort_unstable();
        Self::backtrack(&mut res, &mut tmp, &candidates, candidates.len(), 0, target);

        return res;
    }
}
